glean-core/src/metrics/mod.rs

```diff
@@ -31,37 +31,36 @@ mod uuid;
 
 pub use crate::event_database::RecordedEvent;
 use crate::histogram::{Functional, Histogram, PrecomputedExponential, PrecomputedLinear};
-pub use crate::metrics::datetime::Datetime;
+pub use crate::metrics::datetime::ChronoDatetime;
 use crate::util::get_iso_time_string;
 use crate::CommonMetricData;
 use crate::Glean;
 
-pub use self::boolean::BooleanMetric;
-pub use self::counter::CounterMetric;
-pub use self::datetime::DatetimeMetric;
-pub use self::event::EventMetric;
+pub use self::boolean::{Boolean, BooleanMetric};
+pub use self::counter::{Counter, CounterMetric};
+pub use self::datetime::{Datetime, DatetimeMetric};
+pub use self::event::{Event, EventMetric};
 pub(crate) use self::experiment::ExperimentMetric;
 pub use crate::histogram::HistogramType;
 // Note: only expose RecordedExperimentData to tests in
 // the next line, so that glean-core\src\lib.rs won't fail to build.
-pub use self::custom_distribution::CustomDistributionMetric;
+pub use self::custom_distribution::{CustomDistribution, CustomDistributionMetric};
 #[cfg(test)]
 pub(crate) use self::experiment::RecordedExperimentData;
-pub use self::jwe::JweMetric;
+pub use self::jwe::{Jwe, JweMetric};
 pub use self::labeled::{
     combine_base_identifier_and_label, dynamic_label, strip_label, LabeledMetric,
 };
-pub use self::memory_distribution::MemoryDistributionMetric;
+pub use self::memory_distribution::{MemoryDistribution, MemoryDistributionMetric};
 pub use self::memory_unit::MemoryUnit;
 pub use self::ping::PingType;
-pub use self::quantity::QuantityMetric;
-pub use self::string::StringMetric;
-pub use self::string_list::StringListMetric;
+pub use self::quantity::{Quantity, QuantityMetric};
+pub use self::string::{/*String, THIS FAILS HARD TODO */ StringMetric};
```

Comment:
Now, if we retain the `String` name for the trait (or for `Datetime`, or `Uuid`) then we start colliding with type names from other crates and that makes the build blow up.

Bikeshedding time:

- `StringContract`?
- `StringOperations`?

@badboy @chutten ... thoughts?

Reply:
It's verbose and Java-ey, but maybe `StringMetricTrait`?  `StringMetricRecorder`?

Reply:
> 
> 
> It's verbose and Java-ey, but maybe `StringMetricTrait`? `StringMetricRecorder`?

Jan-Erik vetoed the `Trait` suffix (for good reason, and I agree!). @mdboom what's the idea behind "recorder"?

Reply:
It's the interface that allows "recording" to a metric...  (Just grasping for something...)

Reply:
Thinking my thoughts on this, I think I like having `Metric` in it, since it really has to do with "string metrics" not "strings" in general.  But still don't have a good suggestion for the suffix.

There is the convention to put an `I` in front of interfaces, but that's obviously deeply not Rust like.

Reply:
> 
> 
> It's the interface that allows "recording" to a metric... (Just grasping for something...)

Oh! That's a good point. I'm up for a `StringRecorder` then! Let's see what others think.

Reply:
I think we should use modules to differentiate between metric types, their traits (& libstd's types).

That way `glean_core::metrics::String` can be the type, `glean_core::traits::String` can be the trait.
In code we can `use glean_core::metrics` and `glean_core::traits` and then use the more-qualified path to the type like `metrics::String` instead. That avoids awkward (and redudant) suffixes on type or trait names.

Reply:
(This would mean we create a new directory `glean-core/src/traits/`, add `mod.rs` there and several new `counter.rs`, `string.rs`, etc. Makes looking for just a file name a bit harder, but at least I can still see the path in my editor, so can differentiate them if I open both the metric type and trait.(